### DISCUSSION TITLE
Discussion to Understand TP and CodeGen --> Help make FX graph with torch.compile

### DISCUSSION THREAD

**Discussion by RylieWeaver** [e3nn/e3nn#481] [General]
@mitkotak 

Starting the convo here!

(1) Codegens are for optimizing operations, correct? So, why the distinction between left and right; why have a codegen right but not a left? I am guessing this is just because optimizing operations involving irreps_in1 is just as good as optimizing operations involving irreps_in2, so no reason to have both options?
(2) Internal weights are for contracting all the irreps down to the desired output, right? So, for example, if we had (irreps_in1=8x1o, irreps_in2=8x1o, and irreps_out=8x1e), then we would need to contract down the 64 ways to get a 1e vector given our inputs down to 8?
(3) On the 'mode' given in instructions, 'uvu' would indicate a single weight for all irreps_out that come from an interaction involving irreps_in1{u}, correct? So, for example, if we had (irreps_in1=4x1o, irreps_in2=16x1o, and irreps_out=8x1e), there would be exactly 4 learnable weights?

Looking forward to really understanding it deeply and then getting torch.compile integrated :)

**Comment by RylieWeaver** [comment]
@mitkotak 

Also, random question: Are there pooling operations offered for non-scalar irreps in e3nn?

**Comment by mitkotak** [comment]
Sorry for the delay !

(1) So ``left_right`` is the normal one and ``right`` is for voxel nets but yeah we can stick to the kind of interface that e3nn-jax.FullyConnected has

(2) Yup think of this operation as TP + Linear. The TP takes care of all of the irreps whereas Linear can help upcast or downcast the number of irreps so 64 in this case got broadcasted to 8

(3) Yup uvu is generally used when we are doing a TFN style convolution 128x0e + 128x1o ... \otimes 0e + 1o .... -> 128x0e..... so we ignore the multiplicity from the second dim (typically spherical harmonics)

**Comment by RylieWeaver** [comment]
Hmm, thanks. I'm starting to get a little bit of a picture now. I've been playing around a little in [this](https://colab.research.google.com/drive/1_AfWEep6gmbs8gM5Xo58xgUvxc-kUASO?usp=sharing) colab. According to how I'm understanding this:

(1) The core of how computational graph is created is with operations "graph.___", such as, 'graph.placeholder', etc... In essence, what's done is: (a) Set our x1s, x2s, and weights as an fx.Proxy(). This will make sure all the operations such as einsums are tracked. (b) Do the operations needed for the tensor product (c) Declare the output so the graph has a clear end

(2) We should be able to clear a whole bunch out b calling torch.compile(). All the calls of graph, etc... wil be unnecessary. We can just do our operations, wrap the function, and it will track it all. The place that will be affected most is 'codegen()', followed by 'CodeGenMinIn' and 'TensorProduct'.
